cpp/src/lists/contains.cu

```diff
@@ -44,296 +41,241 @@
+  SearchKeyIter const keys_iter;
+  duplicate_find_option const find_option;
 
-  if constexpr (find_option == duplicate_find_option::FIND_FIRST) {
-    return find_iter - begin;  // Distance of find_position from begin.
-  } else {
-    return end - find_iter - 1;  // Distance of find_position from end.
+  __device__ thrust::pair<size_type, bool> operator()(size_type const idx) const
+  {
+    auto const list = list_device_view{lists, idx};
+    // A null list never contains any key, even null key.
+    // In addition, a null list will result in a null output row.
+    if (list.is_null()) { return {NOT_FOUND_SENTINEL, false}; }
+
+    auto const key_opt = keys_iter[idx];
+    // A null key will also result in a null output row.
+    if (!key_opt) { return {NOT_FOUND_SENTINEL, false}; }
+
+    auto const& key = key_opt.value();
+    return {find_option == duplicate_find_option::FIND_FIRST ? search_list<true>(list, key)
+                                                             : search_list<false>(list, key),
+            true};
   }
-}
 
-/**
- * @brief __device__ functor to search for a key in a `list_device_view`.
- */
-template <duplicate_find_option find_option>
-struct finder {
-  template <typename ElementType>
-  __device__ size_type operator()(list_device_view const& list, ElementType const& search_key) const
+ private:
+  template <bool find_first>
+  static __device__ size_type search_list(list_device_view const& list, Type const& search_key)
   {
-    auto const list_begin = find_begin<ElementType, find_option>(list);
-    auto const list_end   = find_end<ElementType, find_option>(list);
-    auto const find_iter  = thrust::find_if(
-      thrust::seq, list_begin, list_end, [search_key] __device__(auto element_and_validity) {
-        auto [element, element_is_valid] = element_and_validity;
-        return element_is_valid && cudf::equality_compare(element, search_key);
+    auto const [begin, end] = element_index_pair_iter<find_first>(list.size());
+    auto const found_iter =
+      thrust::find_if(thrust::seq, begin, end, [&] __device__(auto const idx) {
+        return !list.is_null(idx) &&
+               cudf::equality_compare(list.template element<Type>(idx), search_key);
       });
-    return distance<find_option>(list_begin, list_end, find_iter);
-  };
+    // If the key is found, return its found position in the list from `found_iter`.
+    return found_iter == end ? NOT_FOUND_SENTINEL : *found_iter;
+  }
 };
 
 /**
- * @brief Functor to search each list row for the specified search keys.
+ * @brief Create a device pointer to the search key(s).
+ *
+ * The returned pointer will be used to construct a validity iterator or an optional iterator for
+ * the keys, which also have the same interface for both `scalar` and `column_device_view`.
+ *
+ * @return Depending on the type of the input key(s), a `scalar` pointer or a `column_device_view`
+ *         pointer will be returned.
```

Comment:
This seems incredibly error prone. In general, a function that differs in the lifetime semantics of the values it returns is not good.

Reply:
Hmm. In fairness, this predates the current change. (i.e. The error is mine.)

The aim of this was to make the lookup code common between the `column_view` and `scalar` paths. There might be a safer way to do this, but it's eluded me thus far.